### ISSUE TITLE
Use router instance in some composable or any other external file

### ISSUE THREAD

**Issue by antony-k1208** [frandiox/vite-ssr#80]
Hi,

maybe I am missing something, but currently the router instance is only getting returned inside the viteSSR HookParams.

What if I want to use some kind of axios interceptor, in order to automatically redirect the user to some route in case of a 401 for example.

Also I am thinking about implementing an auth composable in which there should be a login and logout method. These need to be able to redirect the user to some routes.

How would you build this kind of thing? Do I really have to build everything router related inside the hook function?

Thank your very much in advance for all your work and effort.

**Comment by MarvinRudolph** [comment]
I think if we would add the `router` to the context here https://github.com/frandiox/vite-ssr/blob/master/src/vue/entry-client.ts#L61 it should be available with `useContext()`. What do you think @frandiox, any specific reason to not have it in there?

**Comment by antony-k1208** [comment]
Exactly, that was I was looking for it also. But since it is not available on the context, I was wondering if I am missing something. 

Imho its the only place where it would make sense.

**Comment by frandiox** [comment]
I didn't add it to the context because you can just use vue-router's `useRouter` to get it, no?
https://next.router.vuejs.org/api/#useroute

**Comment by antony-k1208** [comment]
Sadly you can't. At least not everywhere, as mentioned in the docs: `Must be called inside of setup().`

**Comment by antony-k1208** [comment]
I've added a PR for this purpose #81

**Comment by MarvinRudolph** [comment]
Not sure, but afaik `useContext` also needs to be called inside `setup`. Or does it work for you @antony-k1208?

**Comment by frandiox** [comment]
@antony-k1208 As Marving mentioned, `useContext` is also a hook that can only be called in the same places as `useRouter` can. Therefore, I don' think it fixes any situation that `useRouter` doesn't already cover, right?

The reason we need to create/use the Router within a function, instead of exporting it from a different file, is that we need to make sure it is a fresh instance on every SSR request. Otherwise, data would could be mixed from different requests in a Node.js server.

You can implement your own "addCustomHooksToRouter" function that accepts the router instance as a parameter and call it from the vite-ssr main hook or from your app root component.
Wouldn't this work for your use case?

I'm not really against adding the router to the context as a "shortcut" but I thought doing it the Vue way would be better.

**Comment by antony-k1208** [comment]
@frandiox I see. Thanks for your reply and the clarification. So I have to find a solution for my use case =)

Maybe we should merge the PR anyway, since the react part is already working like that. The router instance can be imported via the context. The two implementations should be aligned I'd say. And in react there are also corresponding hooks to get the router stuff or am I missing something? What do you say?

**Comment by frandiox** [comment]
@antony-k1208 Well, the React version works quite differently. It's true that it provides `router` in the context, but that one is not an instance of "React Router". It's a custom router utility that we create in `vite-ssr` because React Router does not provide enough flexibility to await for async routes.

That's why I didn't add the Vue-router instance to the context, because if you get the router there you might think it's a custom wrapper and try to find docs about it. On the other hand, if you use `useRouter` from Vue-router, you can be sure it's a normal Vue-router instance 🤔

**Comment by antony-k1208** [comment]
Okay, thanks for clearing things up. I don't know if users might think, that this is a custom wrapper and will search for some kind of documentation in that regards. I am totally fine with both solutions and learned a thing or two more, so I am happy either way =).

Thanks for your effort.
